GitAutoDeploy.py

```diff
@@ -1,325 +1,504 @@
+
+    def do_POST(self):
+        """Invoked on incoming POST requests"""
+        from threading import Timer
+
+        # Extract repository URL(s) from incoming request body
+        repo_urls = self.get_repo_urls_from_request()
+
+        self.send_response(200)
+        self.send_header('Content-type', 'text/plain')
+        self.end_headers()
+
+        # Wait one second before we do git pull (why?)
+        Timer(1.0, GitAutoDeploy.process_repo_urls, [repo_urls]).start()
+
+    def get_repo_urls_from_request(self):
+        """Parses the incoming request and extracts all possible URLs to the repository in question. Since repos can
+        have both ssh://, git:// and https:// URIs, and we don't know which of them is specified in the config, we need
+        to collect and compare them all."""
+        import json
+
+        #content_type = self.headers.getheader('content-type')
+        length = int(self.headers.getheader('content-length'))
+        body = self.rfile.read(length)
+
+        data = json.loads(body)
+
+        repo_urls = []
+
+        gitlab_event = self.headers.getheader('X-Gitlab-Event')
+        github_event = self.headers.getheader('X-GitHub-Event')
+        user_agent = self.headers.getheader('User-Agent')
+
+        if not 'repository' in data:
+            print "ERROR - Unable to recognize data format"
+            return repo_urls
+
+        # Assume GitLab if the X-Gitlab-Event HTTP header is set
+        if gitlab_event:
+
+            print "Received '%s' event from GitLab" % gitlab_event
+
+            # One repository may posses multiple URLs for different protocols
+            for k in ['url', 'git_http_url', 'git_ssh_url']:
+                if k in data['repository']:
+                    repo_urls.append(data['repository'][k])
+
+        # Assume GitHub if the X-GitHub-Event HTTP header is set
+        elif github_event:
+
+            print "Received '%s' event from GitHub" % github_event
+
+            # One repository may posses multiple URLs for different protocols
+            for k in ['url', 'git_url', 'clone_url', 'ssh_url']:
+                if k in data['repository']:
+                    repo_urls.append(data['repository'][k])
+
+        # Assume BitBucket if the User-Agent HTTP header is set to 'Bitbucket-Webhooks/2.0' (or something similar)
+        elif user_agent.lower().find('bitbucket') != -1:
+
+            print "Received event from BitBucket"
+
+            # One repository may posses multiple URLs for different protocols
+            for k in ['url', 'git_url', 'clone_url', 'ssh_url']:
+                if k in data['repository']:
+                    repo_urls.append(data['repository'][k])
+
+            if 'full_name' in data['repository']:
+                repo_urls.append('git@bitbucket.org:%s.git' % data['repository']['full_name'])
+                repo_urls.append('https://oliverpoignant@bitbucket.org/%s.git' % data['repository']['full_name'])
```

Comment:
Yes you are right. I made a quick fix, but this line will need more attention since Bitbucket in contrast to GitHub and GitLab uses a URL scheme where the username used for authentication should be included in the URL. I'll need to take a second look to this later.